File: backend-nodejs/src/modules/customer/controllers/clientController.ts
import { Request, Response } from 'express';
import { asyncHandler } from '../../../common/middleware/asyncHandler';
import { AppError } from '../../../common/utils/AppError';
import { logger } from '../../../common/utils/logger';
import { clientService } from '../services/clientService';
import { clientExcelService } from '../services/clientExcelService';
import { s3Service } from '../../../common/services/s3Service';
import { ClientStatus } from '../enums/clientEnums';

// 1. GET /clients - List clients with search, status/RM filtering, and cursor pagination
export const getClients = asyncHandler(async (req: Request, res: Response) => {
  const { status, rm_id, search, cursor, page_size } = req.query;

  const pageSize = page_size ? parseInt(page_size as string, 10) : undefined;
  const clientStatus = status ? (status as ClientStatus) : undefined;

  const response = await clientService.getClients({
    status: clientStatus,
    rmId: rm_id as string | undefined,
    search: search as string | undefined,
    cursor: cursor as string | undefined,
    pageSize,
  });

  return res.status(200).json(response);
});

// 2. GET /clients/:id - Get single client by ID
export const getClientById = asyncHandler(async (req: Request, res: Response) => {
  const client = await clientService.getClientById(req.params.id);
  return res.status(200).json(client);
});

// 3. POST /clients - Create a new client
export const createClient = asyncHandler(async (req: Request, res: Response) => {
  const { firstName, lastName, email, phone } = req.body;

  if (!firstName || !lastName || !email || !phone) {
    logger.warn({ ip: req.ip }, 'Client creation failed: Missing required fields');
    throw new AppError('First name, last name, email, and phone are required', 400);
  }

  const currentUserId = req.user?._id ? req.user._id.toString() : undefined;
  const client = await clientService.createClient(req.body, currentUserId);
  return res.status(201).json(client);
});

// 4. PATCH /clients/:id/status - Update client operational status
export const updateClientStatus = asyncHandler(async (req: Request, res: Response) => {
  const { status } = req.body;

  if (!status || !Object.values(ClientStatus).includes(status)) {
    logger.warn({ ip: req.ip, status }, 'Update client status failed: Invalid or missing status');
    throw new AppError('Valid client status is required', 400);
  }

  const client = await clientService.updateClientStatus(req.params.id, req.body);
  return res.status(200).json(client);
});

// 5. PATCH /clients/:id/rm - Update client relationship manager
export const updateClientRm = asyncHandler(async (req: Request, res: Response) => {
  const { rmId } = req.body;

  if (!rmId) {
    logger.warn({ ip: req.ip }, 'Update client RM failed: Missing rmId');
    throw new AppError('Relationship manager ID (rmId) is required', 400);
  }

  const client = await clientService.updateClientRm(req.params.id, req.body);
  return res.status(200).json(client);
});

// 6. GET /clients/:id/profile - Get client KYC & address profile
export const getClientProfile = asyncHandler(async (req: Request, res: Response) => {
  const profile = await clientService.getClientProfile(req.params.id);
  return res.status(200).json(profile);
});

// 7. POST /clients/:id/profile/verify - Verify KYC status (approving promotes to ACTIVE)
export const verifyKyc = asyncHandler(async (req: Request, res: Response) => {
  const { kycStatus } = req.body;

  if (!kycStatus) {
    logger.warn({ ip: req.ip }, 'Verify KYC failed: Missing kycStatus');
    throw new AppError('kycStatus is required', 400);
  }

  const profile = await clientService.verifyKyc(req.params.id, req.body);
  return res.status(200).json(profile);
});

// 8. POST /clients/bulk-reassign - Bulk reassign RM
export const bulkReassignRm = asyncHandler(async (req: Request, res: Response) => {
  const { clientIds, newRmId } = req.body;

  if (!Array.isArray(clientIds) || clientIds.length === 0 || !newRmId) {
    logger.warn({ ip: req.ip }, 'Bulk reassign RM failed: Invalid input payload');
    throw new AppError('clientIds array and newRmId are required', 400);
  }

  const result = await clientService.bulkReassignRm(req.body);
  return res.status(200).json(result);
});

// 9. GET /clients/bulk-template - Download Excel template
export const downloadBulkTemplate = asyncHandler(async (_req: Request, res: Response) => {
  const excelBuffer = await clientExcelService.generateClientBulkTemplate();

  res.setHeader('Content-Disposition', 'attachment; filename="clients_bulk_template.xlsx"');
  res.setHeader(
    'Content-Type',
    'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
  );

  return res.status(200).send(excelBuffer);
});

// 10. POST /clients/bulk-upload - Upload Excel for async ingestion
export const uploadBulkClients = asyncHandler(async (req: Request, res: Response) => {
  if (!req.file) {
    logger.warn({ ip: req.ip }, 'Bulk client upload failed: No file uploaded');
    throw new AppError('Please select an Excel file to upload', 400);
  }

  const originalName = req.file.originalname;
  if (!originalName.endsWith('.xlsx') && !originalName.endsWith('.xls')) {
    logger.warn({ filename: originalName, ip: req.ip }, 'Bulk client upload failed: Invalid file format');
    throw new AppError('Only Excel files (.xlsx, .xls) are supported', 400);
  }

  const currentUserId = req.user?._id ? req.user._id.toString() : undefined;
  const safeFilename = originalName.replace(/[^a-zA-Z0-9._-]/g, '_');
  const s3Key = `client-uploads/${Date.now()}_${safeFilename}`;
  let presignedUrl: string | null = null;

  try {
    await s3Service.uploadFile({
      key: s3Key,
      buffer: req.file.buffer,
      contentType: req.file.mimetype || 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    });
    presignedUrl = await s3Service.getPresignedDownloadUrl(s3Key);
  } catch (s3Err: unknown) {
    const msg = s3Err instanceof Error ? s3Err.message : String(s3Err);
    logger.warn(
      { error: msg, s3Key },
      'S3 upload failed for client bulk upload. Proceeding with async database ingestion.'
    );
  }

  // Dispatch asynchronous background processing
  void clientService.processBulkUploadAsync(req.file.buffer, currentUserId);

  return res.status(202).json({
    status: 'PROCESSING',
    message: 'Bulk client upload is being processed in background',
    filename: originalName,
    s3Key,
    fileUrl: presignedUrl,
  });
});

// 11. GET /clients/bulk-uploads/download-url - Retrieve time-limited pre-signed URL for an uploaded batch file
export const getBulkUploadDownloadUrl = asyncHandler(async (req: Request, res: Response) => {
  const key = typeof req.query.key === 'string' ? req.query.key : undefined;

  if (!key) {
    logger.warn({ ip: req.ip }, 'Get bulk upload download URL failed: Missing key query param');
    throw new AppError('Query parameter "key" is required', 400);
  }

  if (key.includes('..')) {
    logger.warn({ ip: req.ip, key }, 'Get bulk upload download URL rejected: Path traversal detected');
    throw new AppError('Invalid S3 key path', 400);
  }

  const fileUrl = await s3Service.getPresignedDownloadUrl(key);
  return res.status(200).json({ s3Key: key, fileUrl });
});

File: backend-java/src/main/java/com/wealthtech/crm/modules/customer/controller/ClientController.java
package com.wealthtech.crm.modules.customer.controller;

import java.io.IOException;
import java.util.Map;

import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import com.wealthtech.crm.modules.customer.dto.BulkReassignRmRequest;
import com.wealthtech.crm.modules.customer.dto.BulkUploadResponse;
import com.wealthtech.crm.modules.customer.dto.ClientProfileResponse;
import com.wealthtech.crm.modules.customer.dto.ClientResponse;
import com.wealthtech.crm.modules.customer.dto.CreateClientRequest;
import com.wealthtech.crm.modules.customer.dto.UpdateClientRmRequest;
import com.wealthtech.crm.modules.customer.dto.UpdateClientStatusRequest;
import com.wealthtech.crm.modules.customer.dto.VerifyKycRequest;
import com.wealthtech.crm.modules.customer.enums.ClientStatus;
import com.wealthtech.crm.modules.customer.service.ClientExcelService;
import com.wealthtech.crm.modules.customer.service.ClientService;
import com.wealthtech.crm.modules.usermanager.dto.CursorPaginatedResponse;
import com.wealthtech.crm.modules.usermanager.entity.User;
import com.wealthtech.crm.infrastructure.s3.S3Service;
import com.wealthtech.crm.modules.usermanager.exception.BadRequestException;
import com.wealthtech.crm.modules.usermanager.repository.UserRepository;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@RestController
@RequestMapping("/java-wtc-api/v1/clients")
@RequiredArgsConstructor
public class ClientController {

    private final ClientService clientService;
    private final ClientExcelService excelService;
    private final UserRepository userRepository;
    private final S3Service s3Service;

    // List clients with cursor pagination & search/status/RM filtering
    @GetMapping
    @PreAuthorize("hasAuthority('client:read')")
    public ResponseEntity<CursorPaginatedResponse<ClientResponse>> getClients(
            @RequestParam(required = false) ClientStatus status,
            @RequestParam(name = "rm_id", required = false) Long rmId,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) Long cursor,
            @RequestParam(name = "page_size", defaultValue = "50") int pageSize
    ) {
        CursorPaginatedResponse<ClientResponse> response = clientService.getClientList(status, rmId, search, cursor, pageSize);
        return ResponseEntity.ok(response);
    }

    // Get single client by id
    @GetMapping("/{id}")
    @PreAuthorize("hasAuthority('client:read')")
    public ResponseEntity<ClientResponse> getClient(@PathVariable Long id) {
        ClientResponse response = clientService.getClient(id);
        return ResponseEntity.ok(response);
    }

    // Create a new client
    @PostMapping
    @PreAuthorize("hasAuthority('client:create')")
    public ResponseEntity<ClientResponse> createClient(
            @Valid @RequestBody CreateClientRequest request,
            Authentication authentication
    ) {
        Long currentUserId = resolveCurrentUserId(authentication);
        ClientResponse response = clientService.createClient(request, currentUserId);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    // Update client operational status
    @PatchMapping("/{id}/status")
    @PreAuthorize("hasAuthority('client:update')")
    public ResponseEntity<ClientResponse> updateClientStatus(
            @PathVariable Long id,
            @Valid @RequestBody UpdateClientStatusRequest request
    ) {
        ClientResponse response = clientService.updateClientStatus(id, request);
        return ResponseEntity.ok(response);
    }

    // Update client relationship manager
    @PatchMapping("/{id}/rm")
    @PreAuthorize("hasAuthority('client:update')")
    public ResponseEntity<ClientResponse> updateClientRm(
            @PathVariable Long id,
            @Valid @RequestBody UpdateClientRmRequest request
    ) {
        ClientResponse response = clientService.updateClientRm(id, request);
        return ResponseEntity.ok(response);
    }

    // Get client KYC & address profile
    @GetMapping("/{id}/profile")
    @PreAuthorize("hasAuthority('clientprofile:read')")
    public ResponseEntity<ClientProfileResponse> getClientProfile(@PathVariable Long id) {
        ClientProfileResponse response = clientService.getClientProfile(id);
        return ResponseEntity.ok(response);
    }

    // Verify KYC documents (approval or rejection)
    @PostMapping("/{id}/profile/verify")
    @PreAuthorize("hasAuthority('clientprofile:update')")
    public ResponseEntity<ClientProfileResponse> verifyKyc(
            @PathVariable Long id,
            @Valid @RequestBody VerifyKycRequest request
    ) {
        ClientProfileResponse response = clientService.verifyKyc(id, request);
        return ResponseEntity.ok(response);
    }

    // Bulk reassign relationship manager
    @PostMapping("/bulk-reassign")
    @PreAuthorize("hasAuthority('client:update')")
    public ResponseEntity<Map<String, Object>> bulkReassignRm(
            @Valid @RequestBody BulkReassignRmRequest request
    ) {
        int updatedCount = clientService.bulkReassignRm(request);
        return ResponseEntity.ok(Map.of(
                "message", "Successfully reassigned relationship manager",
                "count", updatedCount
        ));
    }

    // Download bulk upload Excel (.xlsx) template
    @GetMapping("/bulk-template")
    @PreAuthorize("hasAuthority('client:create')")
    public ResponseEntity<byte[]> downloadBulkTemplate() {
        byte[] excelBytes = excelService.generateClientBulkTemplate();
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"clients_bulk_template.xlsx\"")
                .contentType(MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"))
                .body(excelBytes);
    }

    // Upload Excel (.xlsx) file for asynchronous batch client ingestion
    @PostMapping(value = "/bulk-upload", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    @PreAuthorize("hasAuthority('client:create')")
    public ResponseEntity<BulkUploadResponse> uploadBulkClients(
            @RequestParam("file") MultipartFile file,
            Authentication authentication
    ) {
        if (file == null || file.isEmpty()) {
            throw new BadRequestException("Please select an Excel file to upload");
        }

        String filename = file.getOriginalFilename();
        if (filename == null || (!filename.endsWith(".xlsx") && !filename.endsWith(".xls"))) {
            throw new BadRequestException("Only Excel files (.xlsx, .xls) are supported");
        }

        Long currentUserId = resolveCurrentUserId(authentication);
        byte[] fileBytes;
        try {
            fileBytes = file.getBytes();
        } catch (IOException e) {
            throw new BadRequestException("Failed to read uploaded file contents");
        }

        String safeFilename = filename.replaceAll("[^a-zA-Z0-9._-]", "_");
        String s3Key = "client-uploads/" + System.currentTimeMillis() + "_" + safeFilename;
        String presignedUrl = null;

        try {
            s3Service.uploadFile(s3Key, fileBytes, file.getContentType());
            presignedUrl = s3Service.generatePresignedGetUrl(s3Key);
        } catch (Exception e) {
            log.warn("S3 upload for client bulk upload failed: {}. Proceeding with async database ingestion.", e.getMessage());
        }

        clientService.processBulkUploadAsync(fileBytes, currentUserId);

        BulkUploadResponse response = new BulkUploadResponse(
                "PROCESSING",
                "Bulk client upload is being processed in background",
                filename,
                s3Key,
                presignedUrl
        );
        return ResponseEntity.status(HttpStatus.ACCEPTED).body(response);
    }

    private Long resolveCurrentUserId(Authentication authentication) {
        if (authentication == null) {
            return null;
        }
        String email = authentication.getName();
        return userRepository.findByEmail(email)
                .map(User::getId)
                .orElse(null);
    }
}
